### PR TITLE
feat(python): Embedding API fine tuning support

### PR DESCRIPTION
**Pull request by AyushExel**
# based on https://github.com/lancedb/lancedb/pull/1023
Very WIP. I'm thinking of merging individual pieces in this feature branch instead of main so we can still review code in pieces and avoid polluting main.

- Adds support for creating corpus from llama-index text-node object (aim to remove this dependency)
- Adds very basic support for LLM api for chat completion, will expand as need arises.
- Add basic universal evaluator
- Add Sentence transformer finetuning support

Known problems:

 - [ ] W&B experiment tracking is not working for sentence transformers

### CODE REVIEW

File: python/python/lancedb/embeddings/fine_tuner/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+Fine-tuning workflow for embeddings consists for the following parts:
+
+### QADataset
+This class is used for managing the data for fine-tuning. It contains the following builder methods:
+```
+- from_llm(
```
Comment (raghavdixit99) [inline python/python/lancedb/embeddings/fine_tuner/README.md:6]: '-' is this a typo or by design?

File: python/python/lancedb/embeddings/fine_tuner/dataset.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    metadata: Dict[str, Any] = {}
+
+    @classmethod
+    def from_chunk(cls, chunk: str, metadata: dict = {}) -> "TextChunk":
+        """Create a SimpleTextChunk from a chunk."""
+        # generate a unique id
+        return cls(text=chunk, id=str(uuid.uuid4()), metadata=metadata)
+
+    @classmethod
+    def from_llama_index_node(cls, node):
```
Comment (raghavdixit99) [inline python/python/lancedb/embeddings/fine_tuner/dataset.py:161]: can we add the expected type for node obj , something like `node: TextNode`

File: python/python/lancedb/embeddings/fine_tuner/dataset.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            queries=dict(zip(queries["id"], queries["query"])),
+            corpus=dict(zip(corpus["id"], corpus["text"])),
+            relevant_docs=dict(zip(relevant_docs["query_id"], relevant_docs["doc_id"])),
+        )
+
+    # generate queries as a convenience function
+    @classmethod
+    def from_llm(
+        cls,
+        nodes: "List[TextChunk]",
```
Comment (raghavdixit99) [inline python/python/lancedb/embeddings/fine_tuner/dataset.py:107]: can we add `from __future__ import annotations` and avoid the string forward references ?
or is it to keep python<3.7/3.10 version compatibility?

**Comment by raghavdixit99** [comment]
@AyushExel  added minor comments, otherwise LGTM

File: python/python/lancedb/embeddings/fine_tuner/README.md (hunk 1 above)
Comment (AyushExel) [reply python/python/lancedb/embeddings/fine_tuner/README.md:6]: For rendering as list

File: python/python/lancedb/embeddings/fine_tuner/dataset.py (hunk 3 above)
Comment (AyushExel) [reply python/python/lancedb/embeddings/fine_tuner/dataset.py:107]: compatibility

File: python/python/lancedb/embeddings/fine_tuner/dataset.py (hunk 2 above)
Comment (AyushExel) [reply python/python/lancedb/embeddings/fine_tuner/dataset.py:161]: sg!

**Comment by tanaymeh** [comment]
@AyushExel LGTM 🚀

File: python/python/lancedb/embeddings/fine_tuner/dataset.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+Given the context information and no prior knowledge.
+generate only questions based on the below query.
+
+You are a Teacher/ Professor. Your task is to setup \
+{num_questions_per_chunk} questions for an upcoming \
+quiz/examination. The questions should be diverse in nature \
+across the document. Restrict the questions to the \
+context information provided."
+"""
```
Comment (sankalp1999) [inline python/python/lancedb/embeddings/fine_tuner/dataset.py:28]: Has this prompt been tested enough? Perhaps add a response example?

Comment (AyushExel) [reply python/python/lancedb/embeddings/fine_tuner/dataset.py:28]: Yeah this is from Llama-index, so ample testing has been done :)

File: python/python/lancedb/embeddings/fine_tuner/README.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    def finetune(self) -> None:
+        """Goes off and does stuff."""
+
+    def helper(self) -> None:
+        """A helper method."""
+        pass
+```
+
+### Embedding API finetuning implementation
+Each embedding API needs to implement `finetune` method in order to support fine-tuning. A vinalla evaluation technique has been implemented in the `BaseEmbedding` class that calculates hit_rate @ `top_k`.
```
Comment (sankalp1999) [inline python/python/lancedb/embeddings/fine_tuner/README.md:69]: *vanilla

File: python/python/lancedb/embeddings/fine_tuner/README.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    llm = Openai()
+    
+    # convert Llama-index TextNode to TextChunk
+    chunks = [TextChunk.from_llama_index_node(node) for node in nodes]
+
+    ds = QADataset.from_llm(chunks, llm)
+    ds.save(name)
+
+# 2. Initialize the embedding model
+model = get_registry().get("sentence-transformers").create()
```
Comment (sankalp1999) [inline python/python/lancedb/embeddings/fine_tuner/README.md:113]: miniLm2 being finetuned here i guess?
can possibly mention in comment

Comment (AyushExel) [reply python/python/lancedb/embeddings/fine_tuner/README.md:113]: yeah, this readme is for internal purpose demo.. Once this is in main, I'll write official docs in a detailed manner

Comment (sankalp1999) [reply python/python/lancedb/embeddings/fine_tuner/README.md:113]: cool

**Comment by AyushExel** [comment]
@changhiskhan @

**Comment by AyushExel** [comment]
@raghavdixit99 @changhiskhan @tanaymeh Merging this in feature branch..Will write the fine tuning code for huggingface in a separate PR and then open another PR to main

**Pull request merged**